pkg/volume/util/operationexecutor/operation_executor.go

```diff
@@ -105,6 +107,28 @@ type OperationExecutor interface {
 	// actual state of the world to reflect that.
 	UnmountDevice(deviceToDetach AttachedVolume, actualStateOfWorld ActualStateOfWorldMounterUpdater, mounter mount.Interface) error
 
+	// MapVolume is used when the volumeMode is 'Block'.
+	// This method creates a symbolic link to the volume from both the pod
+	// specified in volumeToMount and global map path.
+	// Specifically it will:
+	// * Wait for the device to finish attaching (for attachable volumes only).
+	// * Update actual state of world to reflect volume is globally mounted/mapped.
+	// * Map volume to global map path using symbolic link.
+	// * Map the volume to the pod device map path using symbolic link.
+	// * Update actual state of world to reflect volume is mounted/mapped to the pod path.
+	MapVolume(waitForAttachTimeout time.Duration, volumeToMount VolumeToMount, actualStateOfWorld ActualStateOfWorldMounterUpdater) error
```

Comment:
Why are new `MapVolume`/`UnmapVolume` operations needed. Why can't the exist `MountDevice`/`UnmountDevice`/MountVolume`/`UnmountVolume` could be extended to capture block behavior?

Reply:
`MountDevice` mounts in plugin global directory for reference counting. `MapVolume` does something completely different.

Reply:
We have three operations `MountVolume`/`UnmountVolume`/`UnmountDevice` currently.
These operations execute `mount volume`, `unmount volume` and `detach volume` based on `mountpoint` and `mountpoint reference counting` information.
As for the Block Volume case, we uses `symbolic link` and `reference count using symbolic link` since the volume isn't formatted.

As @jsafrane mentioned, operations are completely different between filesystem and block, therefore we define new operations which indicates internal steps.

Filesystem | Purpose
-------------- | -------------
MountVolume | Call plugin's MountDevice() and SetUp()
UnmountVolume | Call plugin's TearDown()
UnmountDevice | Call plugin's UnmountDevice() based on reference count of mountpoint

Block | Purpose
-------------- | -------------
MapVolume | Call plugin's SetUpDevice() and create symlink to plugin dir and pod dir. Create loopbak for volume to avoid silent volume replacement
UnmapVolume | Delete symlink under plugin dir and pod dir
UnmapDevice | Call plugin's TearDownDevice() based on reference count of symlink. Delete loopback